File: pcseg/model/segmentor/fusion/rpvnet/range_lib/range_utils/nn/functional/map_count.py
from torch.autograd import Function

import rangelib_cuda

__all__ = ['map_count']

class MapCountGPU(Function):
    @staticmethod
    def forward(ctx, pxpy, max_bs, h, w):
        '''
        args:
            pxpy: torch.Int32 tensor N X 3 (mean batch_id, px, py)
                px range:0~w 
                py range:0~h
            max_bs: int max batch size
            h：count map height
            w: count map width
        return:
            outs:torch.Int32 tensor B X H X W
                the counts of the pxpy in count map
        '''
        outs = rangelib_cuda.map_count_forward(pxpy.contiguous(), max_bs, h, w)
        return outs

map_count_gpu = MapCountGPU.apply

def map_count(pxpy, max_bs, h, w):
    return map_count_gpu(pxpy, max_bs, h, w)

File: pcseg/model/segmentor/fusion/rpvnet/range_lib/range_utils/src/map_count.cpp
#include <torch/torch.h>
#include <vector>
#include "map_count_gpu.h"

at::Tensor map_count_forward(
    const at::Tensor pxpy, // N * 3 : batch_id,px,py
    const int max_bs,
    const int h, // count map height
    const int w // count map width
)
{
    int N = pxpy.size(0);
    at::Tensor count_map = torch::zeros({max_bs,h,w}, at::device(pxpy.device()).dtype(at::ScalarType::Int));
    // pxpy.print();
    map_count_wrapper(N, h, w, pxpy.data_ptr<int>(), count_map.data_ptr<int>());
    return count_map;
}